### PR TITLE
feat: MorphoLeverageModule

### PR DESCRIPTION
**Pull request by ckoopmann**
New Leverage module to use Morpho as underlying money market for leveraged products.
Generally designed to adhere to the Interface of previous Leverage Modules (Aave / Compound) as closely as possible for easier integration. 

TODO:
- [x] First implementation using a single Morpho Market per Set Token
- [x] First set of tests based on previous tests for AaveLeveragemodule
- [x] Extended integration tests including integration with DIM as well as edge case scenarios

### CODE REVIEW

File: contracts/protocol/integration/lib/MorphoMathLib.sol [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+pragma solidity ^0.6.10;
+
+
+/// @title MathLib
+/// @author Morpho Labs
+/// @notice Library to manage fixed-point arithmetic.
+library MorphoMathLib {
+    /// @dev Returns (`x` * `y`) / `d` rounded down.
+    function mulDivDown(uint256 x, uint256 y, uint256 d) internal pure returns (uint256) {
+        return (x * y) / d;
```
[the remark this answers is not in the document]
Comment (MerlinEgalite) [reply contracts/protocol/integration/lib/MorphoMathLib.sol:11]: I would recommend using it yes, else it could lead to very unexpected behaviors

File: contracts/protocol/integration/lib/MorphoSharesMath.sol [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.6.10;
```
[the remark this answers is not in the document]
Comment (MerlinEgalite) [reply contracts/protocol/integration/lib/MorphoSharesMath.sol:2]: same maybe worth using safe math

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol [hunk 3]
```diff
@@ -0,0 +1,920 @@
...
+        IMorpho.MarketParams memory _marketParams
+    )
+        internal
+        view 
+        returns(uint256 collateralBalance, uint256 borrowBalance, uint256 borrowSharesU256)
+    {
+        // TODO: Review if it is more efficient to generate this id once and save in storage
+        bytes32 marketId = _marketParams.id();
+        ( , , uint128 totalBorrowAssets, uint128 totalBorrowShares, , ) = morpho.market(marketId);
+        ( , uint128 borrowShares, uint128 collateral) = morpho.position(marketId, address(_setToken));
```
Comment (MerlinEgalite) [inline contracts/protocol/modules/v1/MorphoLeverageModule.sol:649]: To get up to date balances you can use this library instead: https://github.com/morpho-org/morpho-blue/blob/3f018087e024538486858e87499a10f6283a9528/src/libraries/periphery/MorphoBalancesLib.sol#L109

Else the balance won't take into account the accrued interest, unless interest has been accrued already by triggering the `accrueInterest` function as done [here](https://github.com/IndexCoop/index-protocol/pull/41/files#r1712614077).

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol [hunk 4]
```diff
@@ -0,0 +1,920 @@
...
+    function _repayBorrow(
+        ISetToken _setToken,
+        IMorpho.MarketParams memory _marketParams,
+        uint256 _notionalQuantity,
+        uint256 _shares
+    )
+        internal
+    {
+        // Only ever set shares or assets to avoid "inconsistent input" error when there is a rounding error
+        if(_shares > 0) {
```
Comment (MerlinEgalite) [inline contracts/protocol/modules/v1/MorphoLeverageModule.sol:718]: Using shares is recommended when you want to fully close a position. It avoids leaving dust on Morpho.

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol [hunk 5]
```diff
@@ -0,0 +1,920 @@
...
+     * @param _setToken               Instance of the SetToken
+     */
+    function sync(ISetToken _setToken) public nonReentrant onlyValidAndInitializedSet(_setToken) {
+        _sync(_setToken);
+    }
+
+    function _sync(ISetToken _setToken) internal {
+        IMorpho.MarketParams memory setMarketParams = marketParams[_setToken];
+        require(setMarketParams.collateralToken != address(0), "Collateral not set");
+        morpho.accrueInterest(setMarketParams);
```
[the remark this answers is not in the document]
Comment (MerlinEgalite) [reply contracts/protocol/modules/v1/MorphoLeverageModule.sol:366]: It should be called whenever you need updated balances from Morpho

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol [hunk 6]
```diff
@@ -0,0 +1,920 @@
...
+    function _validateMarketParams(
+        IMorpho.MarketParams memory _newMarketParams
+    )
+        internal
+        view
+        returns(bytes32 marketId)
+    {
+        marketId = _newMarketParams.id();
+        (uint128 totalSupplyAssets,,,,,) = morpho.market(marketId);
+        require(totalSupplyAssets > 0, "Market not enabled");
```
[the remark this answers is not in the document]
Comment (MerlinEgalite) [reply contracts/protocol/modules/v1/MorphoLeverageModule.sol:908]: For this i would check that `lastUpdate` is not 0, see [here](https://github.com/morpho-org/morpho-blue/blob/3f018087e024538486858e87499a10f6283a9528/src/Morpho.sol#L154).

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol [hunk 7]
```diff
@@ -0,0 +1,920 @@
...
+        IERC20 _component,
+        bool _isEquity
+    )
+        external
+        override
+        onlyModule(_setToken)
+    {
+        // TODO: Check if this old comment from aave version is still relevant here
+        // Check hook not being called for an equity position. If hook is called with equity position and outstanding borrow position 
+        // exists the loan would be taken out twice potentially leading to liquidation
```
Comment (MerlinEgalite) [inline contracts/protocol/modules/v1/MorphoLeverageModule.sol:558]: Can you explain us what this function is supposed to do?

**Review by MerlinEgalite — APPROVED** [review]

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol (hunk 7 above)
Comment (MerlinEgalite) [reply contracts/protocol/modules/v1/MorphoLeverageModule.sol:564]: Very clear thx

**Review by MerlinEgalite — APPROVED** [review]

File: contracts/protocol/modules/v1/MorphoLeverageModule.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (MerlinEgalite) [reply contracts/protocol/modules/v1/MorphoLeverageModule.sol]: Yes exactly. But let's say that you will do an interaction with Morpho in the tx, it might be more gas efficient to accrue the interest first since it will be done anyway during the supply/borrow/repay/withdraw
